src/main/kotlin/org/rust/lang/core/psi/util/RustModExtensions.kt

```diff
@@ -3,75 +3,95 @@ package org.rust.lang.core.psi.util
+            else -> name?.let                       { RustQualifiedName(RustIdNamePart(it), `super`?.canonicalName) }
+        }
 
+/**
+ *  Returns a parent module (`super::` in paths).
+ *
+ *  The parent module may be in the same or other file.
+ *
+ *  Reference:
+ *    https://doc.rust-lang.org/reference.html#paths
+ */
+public val RustModItem.`super`: RustModItem?
+    get() {
+        if (isCrateRoot) return null
 
-private val RustModItem.isCrateRoot: Boolean
-    get() = containingMod == null &&
-        (containingFile.name == "main.rs" || containingFile.name == "lib.rs")
+        val self = this
+        val superInFile = self.containingMod
+        if (superInFile != null)
+            return superInFile
 
-internal val RustModItem.ownsDirectory: Boolean
-    get() =     containingMod != null // any inline nested module owns a directory
-            ||  containingFile.name == MOD_RS
-            ||  isCrateRoot
-
-val RustModItem.modDecls: Collection<RustModDeclItem>
-    get() = PsiTreeUtil.getChildrenOfTypeAsList(this, RustModDeclItem::class.java)
+        return RustModulesIndex.getSuperFor(self)
+    }
 
-val RustModItem.useDeclarations: Collection<RustUseItem>
-    get() = PsiTreeUtil.getChildrenOfTypeAsList(this, RustUseItem::class.java)
+internal val RustModItem.ownsDirectory: Boolean
+    get() = containingMod != null /* Any inline nested module owns a directory */
+        ||  containingFile.name == RustModules.MOD_RS
+        ||  isCrateRoot
 
+internal val RustModItem.ownedDirectory: PsiDirectory?
+    get() {
+        if (!ownsDirectory) return null
 
-sealed class ChildModFile {
-    val mod: RustModItem?
-        get() = when (this) {
-            is Found -> (file as? RustFileImpl)?.mod
-            else -> null
-        }
+        val parent  = containingMod ?: return containingFile.parent
+        val name    = this.name     ?: return null
 
-    class NotFound(val suggestedName: String? = null) : ChildModFile()
-    class Found(val file: PsiFile) : ChildModFile()
-    class Ambiguous(val paths: Collection<PsiFile>) : ChildModFile()
-}
+        return parent.ownedDirectory?.findSubdirectory(name)
+    }
 
-val RustModDeclItem.moduleFile: ChildModFile
+private val RustModItem.isCrateRoot: Boolean
     get() {
-        val parent = containingMod
-        val name = name
-        if (parent == null || name == null || !parent.ownsDirectory) {
-            return ChildModFile.NotFound()
-        }
-
-        val dir = parent.modDir
-        val dirMod = dir?.findSubdirectory(name)?.findFile(MOD_RS)
-        val fileName = "$name.rs"
-        val fileMod = dir?.findFile(fileName)
+        val file = containingFile.virtualFile
+        val containingDir = file.parent
+        return project.getCrateSourceRootFor(file)?.let {
+            it.equals(containingDir) && (containingFile.name == RustModules.MAIN_RS || containingFile.name == RustModules.LIB_RS)
```

Comment:
There are a number of places in current code where we check for `main.rs` and `lib.rs`. This is not exactly good because a cargo package can have more than two roots. iirc, the possible roots are

* library root, (`src/lib.rs` by default, but can be renamed)
* any number of binary roots (`src/main.rs` and `bin/*.rs`)
* a test root (`tests/test.rs`)
* a benchmark root (not sure what is the default path)

This shouldn't be fixed in this PR, it is already quite large, but it is worth keeping in mind.